### ISSUE TITLE
don't use flags to indicate languages

### ISSUE THREAD

**Issue by h01ger** [pretix/pretix#300]
Hi,

please don't spread the bad habit of using flags to indicate languages.

See https://www.w3.org/International/questions/qa-link-lang and http://www.flagsarenotlanguages.com/blog/why-flags-do-not-represent-language/ for explainations why.

Instead, please use the language name in the language itself, so "English / Deutsch" in this case or more generally عربية (Arabiya) Български (Bəlgarski) català česky dansk Ελληνικά (Ellinika) English español Esperanto فارسی (Farsi) français Galego 한국어 (Hangul) Հայերեն (hayeren) hrvatski Indonesia Italiano עברית (ivrit) Lietuvių magyar Nederlands 日本語 (Nihongo) norsk (bokmål) polski Português română Русский (Russkij) slovensky suomi svenska தமிழ் (Tamil) Tiếng Việt Türkçe українська (ukrajins'ka) 中文(简) 中文(HK) 中文(繁) .

To make things worse, the flags dont even have an alt- or title-attribute, so accessibility is also not given.

Thanks for maintaining this useful software! (Besides this little glitch) I love the UI!

**Comment by raphaelm** [comment]
Good point, thanks a lot. I'm totally in favor of changing this in the frontend and will either do so if I get to it or accept a patch that does this.

In the backend, it'll be harder as we use flags to indicate target languages of input fields. We could use shortcodes there, but it would require additional design decisions and effort.

**Comment by raphaelm** [comment]
**Hey!** If you ever wanted to contribute to an Open Source project, but didn't know how, you just found a perfect place to start. We try to write this issue as instructive as possible, but as we're humans we almost certainly forgot to write something important in here. If you want to try to tackle this issue and get stuck at _any_ point, please do not hesitate and ask us, either here on GitHub or via email at support@pretix.eu. We'll be very happy to help! 

We think this issue is very suitable to people who are just starting with Python or Django development or who never contributed to Open Source before. We'd like to ask people with wide experience in these areas to leave this issue to beginners. Thanks! (If you want something more advanced, just drop us a line, we'll find something :wink:)

## How to get started

First of all, you need pretix running locally on your machine. We have a guide on the developer installation [over here](https://docs.pretix.eu/en/latest/development/setup.html). Play around with it and get a feeling for what pretix is and how it works :) Try to find the places in the interface mentioned in this issue where something needs to change. Maybe try to find the same places in the code? Just searching for characteristic words can be a powerful tool to find your way around.

## What this is about

Currently, pretix' frontend shows flags in the top-right corner to represent multiple languages supported.
You can read in the first comment of this issue, why flags are a bad idea and what we should do instead.

## How to do this

The template containing the flags is ``src/pretix/presale/templates/pretixpresale/event/base.html``. Look out there for the following snippet: 
```
                    {% for l in request.event.settings.locales %}
                        <a href="{% url "presale:locale.set" %}?locale={{ l }}&next={{ request.path }}%3F{{ request.META.QUERY_STRING|urlencode }}"
                            class="{% if l == request.LANGUAGE_CODE %}active{% endif %}"
                            ><img src="{% static "pretixbase/img/flags/"|add:l|add:".png" %}"/></a>
                    {% endfor %}
```
and replace the images by the names of the languages. Those are present in ``settings.LANGUAGES`` but probably not accessible from the template context yet. Please tell me if you get stuck there :)

## I want to give it a try!

You're still reading, nice! If you want to try working on this, please comment below, so everyone else knows this is being looked at and can search for a different place to contribute. 

First, [fork](https://github.com/pretix/pretix/fork) the repository to your own GitHub account. Then clone it locally (or [set it as your remote](https://help.github.com/articles/configuring-a-remote-for-a-fork/) if you already cloned it earlier) and prepare your pull requests. There are a number of tutorials on the Pull Request workflow online, for example this [short version](https://guides.github.com/introduction/flow/) by GitHub itself and this [more in-depth tutorial](https://www.digitalocean.com/community/tutorials/how-to-create-a-pull-request-on-github). We're excited to see your pull request! 

And don't forget: If you get stuck, don't hestitate to ask. We bound ourselves to a [Code of Conduct](https://docs.pretix.eu/en/latest/development/contribution/codeofconduct.html) and we're committed to providing a welcoming experience and giving constructive, friendly and useful feedback on your contributions.

**Comment by chrisdamba** [comment]
Just a word of appreciation and gratitude for the welcoming gesture, very helpful and constructive feedback you give on first-time contributors like me; quite motivational and encouraging. Thank you!

**Comment by faheel** [comment]
I'm working on this issue.

**Comment by morrme** [comment]
So right now the link is represented by the image generated by
`<img src="{% static "pretixbase/img/flags/"|add:l|add:".png" %}"/>`

Are you saying that that img tag should be replaced with one that just displays language name? Or do you want the images in that path to be replaced?

**Comment by raphaelm** [comment]
Yes, I think the image tag should be replaced by the name of the language. In the second step, it should be the name of the language *in* the language, which might be a bit harder. Also, we need to somehow make it look nice then :)

**Comment by morrme** [comment]
i may need a little help with navigation, because i can't find any file that has the full form version of the language name...would this need to be something added?

would you ever do something like this (styled to fit the app) 
![language-names-7057347](https://cloud.githubusercontent.com/assets/26514778/24329670/f486f918-11d2-11e7-87e3-67dd5bec2e56.jpg)

**Comment by raphaelm** [comment]
> i may need a little help with navigation, because i can't find any file that has the full form version of the language name...would this need to be something added?

There is the ``LANGUAGES`` list available from ``django.conf.settings`` that has the form

      LANGUAGES = [
           ('en', _('English')),
           ('de', _('German')),
           ('de-informal', _('German (informal)')),
      ]

  I suggest that, for now, you find a way to use this information (you might have to do that in Python code and pass it to the template somehow, this could be done in ``pretix/presale/context.py`` amongst all the other things passed to the template context. Later, we can think of a way to show the languages with the name within their respective language.

> would you ever do something like this (styled to fit the app)

How do you mean this? Yes, this is how the end result should look like -- of course not with the black boxes and probably not this large. Most events will have two or three languages at most, so we can stick to tiny text links in the place where the flags have been before.

**Comment by morrme** [comment]
I am not sure I could correctly implement this in a timely manner.  Hopefully someone with a better understanding of the application will come along and help.

**Comment by anusha94** [comment]
Hi @raphaelm i could give a try :) Can you please assign this to me?

**Comment by raphaelm** [comment]
Hi @anusha94, sadly, GitHub is unable to formally assign issues to non-project-members. But yeah, go ahead and ask if you have any questions! :)

**Comment by raphaelm** [comment]
Hi @anusha94, how's it going? Can I help you in any way?

**Comment by anusha94** [comment]
Hi @raphaelm  Actually I kinda got stuck with the installation itself, will try out one last time on my own and get back to you

**Comment by mfattoru** [comment]
Is this issue still open?
I'm new to the open source community and i would like to understand how  to contribute from now on.

**Comment by raphaelm** [comment]
Yes, this is still open, if @anusha94 is no longer working on it

**Comment by anusha94** [comment]
@mfattoru Please take over this issue. Haven't been able to work on this one :(

**Comment by mfattoru** [comment]
I've installed the code on the machine,i got only two warnings in the generation of the test data:
`/Users/mfattoru/Dropbox/GitHub/pretix/env/lib/python3.6/site-packages/django/db/models/fields/__init__.py:1451: RuntimeWarning: DateTimeField Event.date_from received a naive datetime (2018-09-04 17:00:00) while time zone support is active.
  RuntimeWarning)
/Users/mfattoru/Dropbox/GitHub/pretix/env/lib/python3.6/site-packages/django/db/models/fields/__init__.py:1451: RuntimeWarning: DateTimeField Event.date_to received a naive datetime (2018-09-06 17:00:00) while time zone support is active.
  RuntimeWarning)`

but i don't think this stopped the installation process.
I've started to go around the web application but i'm not sure to what flags the issue is referencing to.

> pretix' frontend shows flags in the top-right corner

In what page is it showed this flag list? I can't find it.

**Comment by raphaelm** [comment]
> In what page is it showed this flag list? I can't find it.

Create an event, select multiple languages during event creation. Then, click "go to shop"  in the top bar.

**Comment by mfattoru** [comment]
Sorry for the late reply,was really busy during this time.
I was able to find the where to edit the code.
I've found the piece of code to edit and being able to pass the dictionary with different languages to the template.
now i have to access the values inside the dictionary.
For what i understood i need to create a custom filter to access the data inside the templace,considering that i cannot access the value directly from a template.
where exactly should i insert the filter.
the code will be something like:
#inside the filter file:

`
@register.filter
def get_item(dictionary, key):
    return dictionary.get(key)
`
#Inside the template
`{{ languages|get_item:l }}`
where languages is the dictionary containing all the languages,and l is the variable that actually contains the different encoding of the language,so en,de,de-informal etc...

**Comment by mfattoru** [comment]
I've figured out a way to insert a filter,I've added a new python script to the folder src/pretix/base/templatetags.
this is the final result i get from it.

![screenshot-20170703160239-1188x320](https://user-images.githubusercontent.com/20736338/27807895-9a9dca3c-6009-11e7-80e1-14befedb9f6b.png)

considering i'm new to this,what i'm supposed to do now?

**Comment by raphaelm** [comment]
Hi,

I'm not really sure why you think you need a filter. You could just do the following:
```
ctx['languages'] = settings.LANGUAGES
```
and then change the iteration in the template to
```
{% for l, langname in languages %}
    {% if l in request.event.settings.locales %}
        <a …
    {% endif %}
{% endfor %}
```
That would be clearer to read, I guess.

But yes, the end result is the correct one.

**Comment by mfattoru** [comment]
should i change the code in the way you you did it? so not using the filter?
I always thought that is better to put the logic in python scripts and leave the templates as clean as possible,but still i'm way far from being an expert of Django,it's just what they recommended me.
Any tips is really appreciated.
Thank You

**Comment by raphaelm** [comment]
> I always thought that is better to put the logic in python scripts and leave the templates as clean as possible,but still i'm way far from being an expert of Django,it's just what they recommended me.

You are right. However, the filter is also sub-optimal design, since it makes the flow of code harder to understand, because it goes "Python → Template → Python" without necessity. Then, what do you think of the following:

```
if hasattr(request, 'event'):  # This if already exists, just move the next line at the correct place
    ctx['languages'] = [l for l in settings.LANGUAGES if l[0] in request.event.settings.locales]
```
and then just iterate over this variable in the template.

**Comment by mfattoru** [comment]
I like the last implementation

**Comment by raphaelm** [comment]
> I like the last implementation

perfect, then we are on the same page :D Go ahead and create a pull request with it, if you like :)

**Comment by raphaelm** [comment]
Ah, bonus points if you use https://docs.djangoproject.com/en/1.11/topics/i18n/translation/#get-language-info-list or https://docs.djangoproject.com/en/1.11/topics/i18n/translation/#django.utils.translation.get_language_info to display the names of a language *in* the language (e.g. "Deutsch" instead of "German"). But that's not absolutely required for now

**Comment by mfattoru** [comment]
I've completed it and created a the pull request.

**Comment by h01ger** [comment]
On Thu, Jul 06, 2017 at 12:14:00AM -0700, Raphael Michel wrote:
> Closed #300 via 5932558ca29b752aaf5e3634d4f66adbe2d6a037.

many thanks to everyone involved in fixing this!

:)


-- 
cheers,
	Holger
